File: app/src/main/java/com/example/lovereminder/DiaryActivity.java
package com.example.lovereminder;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.lovereminder.databinding.ActivityDiaryBinding;

import org.jetbrains.annotations.NotNull;

import java.text.SimpleDateFormat;
import java.util.Date;

import io.reactivex.CompletableObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class DiaryActivity extends AppCompatActivity{
    private TextView tv_Date;
    private TextView tv_Content;
    private EditText edt_Content;
    private Menu menu;

    private int diaryId;
    String originalText;
    private Diary currentDiary;
    private DiaryViewModel viewModel;
    private DiaryDao mDiaryDao;
    private boolean isEditing;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityDiaryBinding binding = ActivityDiaryBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        mDiaryDao = AppDatabase.getInstance(this).getDiaryDao();
        viewModel = new ViewModelProvider(this).get(DiaryViewModel.class);

        tv_Content = binding.tvContent;
        tv_Date = binding.tvDate;
        edt_Content = binding.edtContent;
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar supportActionBar = getSupportActionBar();
        supportActionBar.setDisplayShowTitleEnabled(false);
        supportActionBar.setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationOnClickListener(view -> onBackPressed());

        if (savedInstanceState != null){
            isEditing = savedInstanceState.getBoolean("isEditing");
            if (isEditing){
                edt_Content.setVisibility(View.VISIBLE);
                tv_Content.setVisibility(View.GONE);
            }
        }
        if (getIntent().hasExtra("id")){
            diaryId = getIntent().getIntExtra("id", 0);
        }

        viewModel.setDiaryId(diaryId);
        viewModel.getDiary().observe(this, diary -> {
            currentDiary = diary;
            originalText = currentDiary.getContent();
            tv_Content.setText(diary.getContent());
            SimpleDateFormat sdf = new SimpleDateFormat("'ngày' dd 'tháng' MM 'năm' yyyy");
            String formattedDateString = sdf.format(new Date(diary.getDate()));
            tv_Date.setText(formattedDateString);
        });
    }

    private void showMessage() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("Chưa lưu thay dổi. Bạn có muốn lưu lại không?")
                .setPositiveButton("Có", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        updateDiary();
                        finish();
                    }
                })
                .setNegativeButton("Không", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        DiaryActivity.super.onBackPressed();
                    }
                });
        AlertDialog alertDialog = builder.create();
        alertDialog.show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the app bar.
        getMenuInflater().inflate(R.menu.menu_edit_diary, menu);
        this.menu = menu;
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean("isEditing", isEditing);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (isEditing){
            menu.findItem(R.id.action_edit_dairy).setVisible(false);
            menu.setGroupVisible(R.id.group_rewrite_diary_status, true);
        }
        else{
            menu.findItem(R.id.action_edit_dairy).setVisible(true);
            menu.setGroupVisible(R.id.group_rewrite_diary_status, false);
        }
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_edit_dairy:
                isEditing = true;

                edt_Content.setVisibility(View.VISIBLE);
                edt_Content.setText(tv_Content.getText().toString());
                edt_Content.requestFocus();
                InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.showSoftInput(edt_Content, InputMethodManager.SHOW_IMPLICIT);
                edt_Content.setSelection(edt_Content.getText().length());
                tv_Content.setVisibility(View.INVISIBLE);

                invalidateOptionsMenu();
                return true;
            case R.id.action_done_rewrite_diary:
                if (originalText.equals(edt_Content.getText().toString().trim())){
                    Toast.makeText(this, "Bạn chưa thay đổi nội dung nhật kí", Toast.LENGTH_SHORT).show();
                    return  true;
                }
                isEditing = false;
                tv_Content.setVisibility(View.VISIBLE);
                edt_Content.setVisibility(View.INVISIBLE);

                updateDiary();

                invalidateOptionsMenu();
                return true;
            case R.id.action_cancel_rewrite_diary:
                isEditing = false;
                tv_Content.setVisibility(View.VISIBLE);
                edt_Content.setVisibility(View.INVISIBLE);
                invalidateOptionsMenu();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void updateDiary() {
        currentDiary.setContent(edt_Content.getText().toString().trim());
//        new UpdateDiaryAsync(mDiaryDao).execute(currentDiary);
        mDiaryDao.updateDiary(currentDiary)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CompletableObserver() {
                    @Override
                    public void onSubscribe(@NotNull Disposable d) {

                    }

                    @Override
                    public void onComplete() {
                        Toast.makeText(DiaryActivity.this, "Diary updated", Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onError(@NotNull Throwable e) {

                    }
                });
    }

    @Override
    public void onBackPressed() {
        if (!originalText.equals(edt_Content.getText().toString().trim()) && isEditing)
            showMessage();
        else super.onBackPressed();
    }

//    private static class UpdateDiaryAsync extends AsyncTask<Diary, Void, Integer>{
//        private DiaryDao mDiaryDao;
//
//        public UpdateDiaryAsync(DiaryDao diaryDao) {
//            mDiaryDao = diaryDao;
//        }
//
//        @Override
//        protected Integer doInBackground(Diary... diaries) {
//            return mDiaryDao.updateDiary(diaries[0]);
//        }
//
//        @Override
//        protected void onPostExecute(Integer updatedDiaryId) {
//            super.onPostExecute(updatedDiaryId);
//            Log.d("Tag", String.valueOf(updatedDiaryId));
//        }
//    }
}

File: app/src/main/java/com/phucnguyen/lovereminder/ui/activity/DiaryActivity.kt
package com.phucnguyen.lovereminder.ui.activity

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.phucnguyen.lovereminder.R
import com.phucnguyen.lovereminder.databinding.ActivityDiaryBinding
import com.phucnguyen.lovereminder.model.Diary
import com.phucnguyen.lovereminder.viewmodel.DiaryViewModel
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

class DiaryActivity : BaseActivity() {
    private var menu: Menu? = null
    private var diaryId = 0
    private lateinit var viewModel: DiaryViewModel
    private lateinit var binding: ActivityDiaryBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme()
        binding = ActivityDiaryBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this).get(DiaryViewModel::class.java)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val supportActionBar = supportActionBar
        supportActionBar!!.setDisplayShowTitleEnabled(false)
        supportActionBar.setDisplayHomeAsUpEnabled(true)
        toolbar.setNavigationOnClickListener { view: View? -> onBackPressed() }
        if (intent.hasExtra(EXTRA_DIARY_ID)) {
            diaryId = intent.getIntExtra(EXTRA_DIARY_ID, DEFAULT_DIARY_ID)
        }

        lifecycleScope.launch { 
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.isEditingFlow.collect {
                        if (it) {
                            binding.edtContent.visibility = View.VISIBLE
                            binding.tvContent.visibility = View.GONE
                            binding.edtContent.setText(binding.tvContent.text.toString())
                            binding.edtContent.requestFocus()
                            val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                            imm.showSoftInput(binding.edtContent, InputMethodManager.SHOW_IMPLICIT)
                            binding.edtContent.setSelection(binding.edtContent.text.length)
                        } else {
                            binding.edtContent.visibility = View.GONE
                            binding.tvContent.visibility = View.VISIBLE
                        }
                    }
                }
                launch {
                    viewModel.findDiaryById(diaryId).collect {
                        populateDiaryView(it)
                    }
                }
            }
        }
    }

    private fun populateDiaryView(diary: Diary) {
        binding.tvContent.text = diary.content
        val sdf = SimpleDateFormat("'ngày' dd 'tháng' MM 'năm' yyyy")
        val formattedDateString = sdf.format(Date(diary.date))
        binding.tvDate.text = formattedDateString
    }

    private fun showMessage() {
        val builder = AlertDialog.Builder(this)
        builder.setMessage(getString(R.string.title_dialog_change_unsave_confirmation))
            .setPositiveButton(getString(R.string.txt_dialog_action_positive)) { dialog, which ->
                lifecycleScope.launch {
                    updateDiary()
                    dialog.dismiss()
                    finish()
                }
            }
            .setNegativeButton(getString(R.string.txt_dialog_action_negative)) { dialog, which ->
                dialog.dismiss()
                super@DiaryActivity.onBackPressed()
            }
        val alertDialog = builder.create()
        alertDialog.show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the app bar.
        menuInflater.inflate(R.menu.menu_edit_diary, menu)
        this.menu = menu
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        if (viewModel.isEditingDiary()) {
            menu.findItem(R.id.action_edit_dairy).isVisible = false
            menu.setGroupVisible(R.id.group_rewrite_diary_status, true)
        } else {
            menu.findItem(R.id.action_edit_dairy).isVisible = true
            menu.setGroupVisible(R.id.group_rewrite_diary_status, false)
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_edit_dairy -> {
                viewModel.setDiaryEditingState(true)
                invalidateOptionsMenu()
                true
            }
            R.id.action_done_rewrite_diary -> {
                if (!viewModel.isDiaryContentChanged(binding.edtContent.text.toString().trim { it <= ' ' })) {
                    Toast.makeText(this, getString(R.string.msg_toast_diary_content_unchanged), Toast.LENGTH_SHORT)
                        .show()
                    return true
                }
                viewModel.setDiaryEditingState(false)
                lifecycleScope.launch {
                    updateDiary()
                }
                invalidateOptionsMenu()
                true
            }
            R.id.action_cancel_rewrite_diary -> {
                viewModel.setDiaryEditingState(false)
                invalidateOptionsMenu()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private suspend fun updateDiary() {
        val newContent = binding.edtContent.text.toString().trim { it <= ' ' }
        val updateResult = viewModel.updateDiary(newContent)
        if (updateResult) {
            Toast.makeText(this@DiaryActivity, getString(R.string.toast_msg_update_diary_success), Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(this@DiaryActivity, getString(R.string.toast_msg_update_diary_failed), Toast.LENGTH_SHORT).show()
        }
    }

    override fun onBackPressed() {
        if (viewModel.isDiaryContentChanged(binding.edtContent.text.toString().trim { it <= ' ' } )
                && viewModel.isEditingDiary()
        ) {
            showMessage()
        }
        else {
            super.onBackPressed()
        }
    }

    companion object {
        const val EXTRA_DIARY_ID = "id"
        const val DEFAULT_DIARY_ID = 0
    }
}